### DISCUSSION TITLE
Snackbar com comportamento persistente

### DISCUSSION THREAD

**Discussion by annefranca** [Telefonica/mistica-design#1467] [Proposals]
### Goal

Avaliar a proposta de comportamento do componente Snackbar para que seja persistente.

### Context

Uns dos princípios que temos na construção do Novo App é a Acessibilidade, com isso, solicitamos que o Snackbar tenha a opção de persistência quando falamos da interação com a pessoa usuária. Acreditamos que não é apenas uma questão visual, mas sim, cultural.

### Impact

Temos definido em nosso conceito deixar a ação sempre visível para a pessoa usuária poder fechar a mensagem e não ter um temporizador, pois isso garante que a pessoa usuária sempre visualize a mensagem por completo.   

Portanto, o Snackbar/Crouton só desaparece da tela após o “tap” da pessoa usuária na ação "Fechar" ou “Entendi” – que são os únicos copys de buttom link permitidos.

### Discussion Questions

ND

**Comment by aweell** [comment]
### Initial considerations

* The ideal behavior of this component is to appear for a limited time and provide contextual information about actions the user recently performed (Example: The user edits the profile information, and the snackbar provides the feedback that the profile information has been successfully updated).

This is aligned with the [Material design usage recommendations for this component](https://m3.material.io/components/snackbar/guidelines#4259e22d-2e3e-4efa-8f56-4d49eea22ddf)

* There are cases where a [non-modal](https://www.nngroup.com/articles/modal-nonmodal-dialog/) dialog element that requires user action could be a solution for displaying low-priority messages to users. (Example: A message that tells the user that third-party cookies are used on the site). This solution is less intrusive than a modal dialog but still is, obscures part of the content until the user decides to dismiss it.

Material 3 recently added:

> Snackbars can either appear temporarily (dismissive) or persist until the user takes an action (non-dismissive)

There are no references yet in the usage documentation about when to use dismissive vs. non-dismissive Snackbars but in material 2, the non-dismissive behaviour was used only [banners](https://m2.material.io/components/banners#usage).

![Screenshot 2023-09-27 at 16 17 56](https://github.com/Telefonica/mistica-design/assets/44420072/27964832-387d-475b-91cb-444ed7b4994e)


We find a symmetry between the banner proposal of material and what Mística has named [Callouts](https://brandfactory.telefonica.com/d/iSp7b1DkYygv/n-a#/components/callout).


### A11y

The previous ideal behavior aligns with [SC 4.1.3 - Status messages (Level AA)](https://www.w3.org/WAI/WCAG22/Understanding/status-messages) where is stated:

> After a user puts a photo in an album in an online photo app, a snackbar displays the message "Saved in 'Wedding' album", which is also read by a screen reader.

Since the current behavior is time-limited, we should then see if we are failing [SC 2.2.3 - No Timing (Level AAA)](https://www.w3.org/WAI/WCAG22/Understanding/no-timing.html)

The current technique ([G5](https://www.w3.org/WAI/WCAG22/Techniques/general/G5)) says:

> The objective of this technique is to provide users with all the time they need to complete an activity.

* Does the user need this information to complete a given task or is complementary?
* Does the user need to dismiss or confirm to complete a given task?
* Is this information available in another place where the user can access it without a time limitation?

It would be very valuable for us to have examples of the use cases that are leading to this requirement.

### Other a11y aspects

[Other aspects related mostly to assistive technologies](https://www.w3.org/WAI/ARIA/apg/patterns/alert/examples/alert/) we agree that should be addressed.

Regarding other types of actions (different than dismiss or confirm) We are currently planning to [allow actions beyond the current onClose of the component](https://github.com/Telefonica/mistica-design/issues/1461), with this new feature we should also review SC 2.2.3 to be compliant.

### Snackbar stacking

The current native behavior regarding the Snackbar staking is that when a second Snackbar is opened the one that is currently displayed disappears.

The unexpected behaviors should be addressed in the UX definition.

### Additional questions

* If the user moves to a different step in the flow or a different screen in the app is the information still relevant?
* Could the actions related to this information change the current position of the user in the app?


### Initial proposal

To address the problems related to the usage of actions in the snackbars, no matter if they're related to the dismissal or not, we propose the following:

* A **snackbar without actions** follows the current definition, after a time delay the snackbar is automatically dismissed
* A **snackbar with actions** can be automatically dismissed or set to permanent. The permanent option should be used in cases where user action is required, for optional or nonrequired actions the auto-dismiss solution could be the way to go.

**Reply by annefranca** [reply]
@aweell 

**Sobre as suas perguntas adicionais:**

Como alinhado via chat no dia 01/08, ao atualizar a página ou sair do fluxo, o componente vai desaparecer, mesmo sem a ação da pessoa usuária.

**Sobre a proposta inicial:** 

Vamos conseguir deixar o Snackbar permanente quando a ação da pessoa usuária é necessária? Se sim, podemos seguir, pois sempre vamos ter uma ação de "Entendi" ou "Fechar" em nossas aplicações.

**Reply by aweell** [reply]
Hey @annefranca 

If you identified your action as required, the snackbar should be able to be configured to be displayed permanently.

We've created #1476 with the proposed changes.

**Comment by aweell** [comment]
I reopen this discussion since there are some considerations regarding latest snackbar behaviour changes:

The [last updated snackbar behaviour](https://www.figma.com/file/yCHLIfy4WMfRdlADwyL4kZ/%F0%9F%94%B8-Snackbar-Specs?type=design&node-id=1038%3A1306&mode=design&t=MfvryvfWCSe1G15a-1) defines:

> When an action is present the snackbar duration can be configured as custom or permanent

But what happens when the action is not related to the dismiss of the snackbar?

In https://github.com/Telefonica/mistica-design/issues/1461 we identified cases where the snackbar action should not take effect when the snackbar is dismissed. 

You can see the example of the current behaviour here -> [Example](https://mistica-web.vercel.app/playroom#?code=N4Igxg9gJgpiBcJgApkEoAEBeAfB4AOgHYYaREDOALhgNpECuAtgJJUxMUA0GFMVAOWZsOFALrYMAJRgBDMFQB0DPgGUqs9sgBMaANzFS5anQBuASwrmARgBsYPPlQBqlm-YlZpchcrUatADNZWz59YkMyCEoaCgALCAB3VSJ5AGsAIVkAJ0l0bDxCElJeflcrOxhkKmyGGHDigF8DIkjjGjjZIih7AEEoKBEmPMxcfEjSJyFWdk5kRhnRDABqDABGBpLeBOTUsEyc9BbSZoji9oxO7vsZJghTGCGRgvHiyf5poYp54VmKDAAtOtNiUnOV3FVgqF6scMKdWsVsvwGNkSMgJhgADw4DGkTEABSRFhgiQAKhAIKEcJiAPSEmDEskUqm4jCszEyCgAB2iVgeABlZABPCAMKg4t5bLEZCAADwwXNkA3MRAA5gBNLDAAAsAA5GhKpSVMep0rxFWAYFrtNqDaytpjSeYqPZtPapdlzKq4lQte6jZiMmKqNF%2BSq0hhovSKBQtVcejB%2BoNZo0MJUYFB1Wn7BmpF6fYajUXSEmMEQSRhnRx-VLaUGqCGiGGiGlC8WTjW2%2B2w9RZDXaU6XTBtF2AzLZbmkj3xTXSMBetlssLFIFshAmMhgBh7GqqHF4GXfktGjxkAB9HjmKCy0Z4dGS9vSuWTxKzqWyGP8P0Px-GlhMWRVRgN9iwobIwCwAgQB9KguQoeAaRpChRXAmBlEoLlbA-OJFEgJgaQATgABiI2ViKImkAH4sO6CgwFkLlgJAECiziGB819YBdSIxoWKND9GIUKRNHMCBIJANZ4DWKC%2BJKGlR1-Xif0fKhnXscTB3sGTlKLeS3zQNAlMfOtnygKQp0sGcdNpU19gUvEaXHBTaU5HlKHMAVhVFcVCBANkQCMqVgAsCp7AwAAycKMHvYsTT2NJrByNkdNIKAURE6JrQAVhIwLi2iABhWwID4OMugTW57keFMa2sYNolJGBZV9KCCq6S1bG0x8mBgGNAKtKCniVWAoAwEMxrY7dLK63SFMMjFaULBpGjQdAlJALgQD3DheoQWgQFcUwIDPAQSRAMRNsSK89woPaAHYADZdTERogA)

And since the duration can be set to permanent, the user should have a way to dismiss the snackbar. So now we have four different scenarios:

1. Snackbar with limited duration and an action 
2. Snackbar with limited duration and dismiss 
3. Snackbar with permanent duration and an action and dismiss
4. Snackbar with permanent duration and dismiss 

The cases you're presenting seem to fall into the fourth definition, permanent duration and dismiss action.

Our proposal here is that when a snackbar is set to dismissable the standard way to dismiss it should be the close icon (same pattern that we are using in other components) so could be not be possible to define a "visible" label for dismiss action like "Fechar" ou “Entendi”, only the text for the action can be defined. 

<img width="404" alt="Screenshot 2023-10-09 at 12 24 16" src="https://github.com/Telefonica/mistica-design/assets/44420072/42b1e2c8-d390-41a7-8f33-9dc5edf4b1f2">

_Example of dismissable snackbar_

For the cases that action and dismiss action are present:

<img width="424" alt="Screenshot 2023-10-09 at 12 06 33" src="https://github.com/Telefonica/mistica-design/assets/44420072/2d5aeca1-cb4f-4ab1-9287-a65b9faa6948">

_Examples of action + dismiss_

**Reply by annefranca** [reply]
@aweell 

Nesse caso apenas o icon-close teria a ação de fechar o componente? A ação padrão que usamos de "Entendi" ou "Fechar" não iria fazer a Snackbar desaparecer?

**Reply by aweell** [reply]
Hey @annefranca 

We are analysing this with development teams in order to find a solution that allows both scenarios. When I have more information I'll come back here.

**Reply by annefranca** [reply]
@aweell perguntei mais para confirmar mesmo. Não seria um problema pra gente a ação de fazer o componente "desaparecer" ser apenas no icon-close. Podemos seguir da forma que você descreveu acima.

**Comment by aweell** [comment]
I close the discussion sin both scenarios will be available in the component. The usage of the action as dismiss is not recommended by design core since could lead to unexpected behaviours in future updates of the component, but at the moment, that would work as intended.
